mm2src/mm2_bitcoin/keys/src/lib.rs

```diff
@@ -38,56 +38,120 @@ pub use signature::{CompactSignature, Signature};
 use hash::{H160, H256};
 use lazy_static::lazy_static;
 use secp256k1::{Secp256k1, SignOnly, VerifyOnly};
-use std::fmt;
+use std::{fmt, hash::Hash};
 
 /// 32 bytes long secret key
 pub type Secret = H256;
 /// 32 bytes long signable message
 pub type Message = H256;
 
-#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+#[derive(Clone, Debug, Eq)]
 pub enum AddressHashEnum {
-    /// 20 bytes long hash derived from public `ripemd160(sha256(public))` used in P2PKH, P2SH, P2WPKH
+    /// 33 or 65 bytes long public key used in P2PKH, P2WPKH, and deprecated P2PK
+    PubKey(Public),
+    /// 20 bytes long hash derived from public `ripemd160(sha256(public))` used in P2PKH, P2WPKH (for when the public key isn't known)
     AddressHash(H160),
+    /// 20 bytes long hash derived from script `ripemd160(sha256(script))` used in P2SH
+    ScriptHash(H160),
```

Comment:
I started with `PubKey` and assumed P2PK, P2PKH, and P2WPKH are all included in this variant. That's why I changed the `AddressHash` name to `ScriptHash` because it only included P2SH types at that moment.

I found later that sometimes we don't have the public key form of a P2PKH or P2WPKH, so instead of putting them in `ScriptHash` I added back `AddressHash` to hold the hash in these cases.

We could merge `AddressHash` and `ScriptHash` variants but I find it more explicit to have them separate.

Reply:
> I found later that sometimes we don't have the public key form of a P2PKH or P2WPKH, so instead of putting them in ScriptHash I added back AddressHash to hold the hash in these cases.

Would adding the Pubkey as optional to `Address` struct instead of adding the new variant `PubKey(Public),` here make more sense. What I see as not logical and confusing is to have `AddressHashEnum::AddressHash(H160)` and not use it for `P2PKH`, what we are aiming to achieve is to access the pubkey easily from `Address` if it's available (it's available for our own addresses only). You can give this solution a try locally and see if it introduces more complexities or makes things simpler.